[threaded pg] make exception handling consistent with MultiProcessTestCase

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90713
* __->__ #90712
* #90595

Differential Revision: [D42153661](https://our.internmc.facebook.com/intern/diff/D42153661)

XilunWu committed:
[threaded pg] make exception handling consistent with MultiProcessTestCase

[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] make exception handling consistent with MultiProcessTestCase"




[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] make exception handling consistent with MultiProcessTestCase"




[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] make exception handling consistent with MultiProcessTestCase"




[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] make exception handling consistent with MultiProcessTestCase"




[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42153661).

wanchaol commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -937,8 +928,8 @@ def __init__(self, method_name: str = "runTest") -> None:
         try:
-            _run_test_with_mt_pg(
-                self=self,
+            self.__class__._run_test_with_mt_pg(
```
nit: maybe just `MultithreadedTestCase._run_test_with_mt_pg`?

wanchaol commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -952,6 +943,65 @@ def perThreadSetUp(self):
+                if skip_code < 0:
+                    skip_code = TEST_SKIPS["generic"].exit_code
+            elif isinstance(exc, TimeoutError):
+                msg = "Thread {} terminated or timed out after {} seconds\n".format(
```
is there going to be other time out messages? shall we just make this the default?

XilunWu replied:
This is the only place to draft a timeout message. Do you mean that we have something like `error_msg = "...".format()` at the beginning of this function?

wanchaol commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -952,6 +943,65 @@ def perThreadSetUp(self):
+
+        # check exceptions
+        if len(error_msg) > 0:
+            raise RuntimeError(error_msg)
```
I think we should re-throw the exception instead of raise a runtime error?

wanchaol replied:
I see, yeah this make sense to me.

wanchaol commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -952,6 +943,65 @@ def perThreadSetUp(self):
+        if len(error_msg) > 0:
+            raise RuntimeError(error_msg)
+        if len(timeout_msg) > 0:
+            raise RuntimeError(timeout_msg)
```
Same here, we should directly raise a TimeoutError

XilunWu replied:
Same as above, MultiProcessTestCase only throws `RuntimeError` and I think we'd better follow the same way. What do you think?

wanchaol replied:
Yeah then the behavior of these two are different, let's just directly throw the TimeoutError whenever TimeoutError got captured.

XilunWu committed:
Update on "[threaded pg] make exception handling consistent with MultiProcessTestCase"


Differential Revision: [D42153661](https://our.internmc.facebook.com/intern/diff/D42153661)

[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] make exception handling consistent with MultiProcessTestCase"


Differential Revision: [D42153661](https://our.internmc.facebook.com/intern/diff/D42153661)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good, please address the comment and fix lint before landing.

XilunWu committed:
Update on "[threaded pg] make exception handling consistent with MultiProcessTestCase"


Differential Revision: [D42153661](https://our.internmc.facebook.com/intern/diff/D42153661)

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42153661).

XilunWu commented:
@pytorchmergebot merge -g
